File: 2025-09-September-GFG-POTD/September 17 - Decode the string/Solution.py
class Solution:
    @staticmethod
    def decodeString(s: str) -> str:
        count_stack = []
        str_stack = []
        curr_str = []
        num = 0

        for c in s:
            if c.isdigit():
                num = num * 10 + int(c)  # handle multi-digit numbers
            elif c == '[':
                count_stack.append(num)
                str_stack.append("".join(curr_str))
                curr_str = []
                num = 0
            elif c == ']':
                repeat = count_stack.pop()
                prev_str = str_stack.pop()
                curr_str = [prev_str + "".join(curr_str) * repeat]
            else:
                curr_str.append(c)

        return "".join(curr_str)

File: 2025-09-September-GFG-POTD/September 17 - Decode the string/Solution.java
class Solution {
    static String decodeString(String s) {
        Stack<Integer> countStack = new Stack<>();
        Stack<StringBuilder> strStack = new Stack<>();
        StringBuilder currStr = new StringBuilder();
        int num = 0;

        for (char c : s.toCharArray()) {
            if (Character.isDigit(c)) {
                num = num * 10 + (c - '0');  // handle multi-digit numbers
            } else if (c == '[') {
                countStack.push(num);
                strStack.push(currStr);
                currStr = new StringBuilder();
                num = 0;
            } else if (c == ']') {
                int repeat = countStack.pop();
                StringBuilder prevStr = strStack.pop();
                for (int i = 0; i < repeat; i++) {
                    prevStr.append(currStr);
                }
                currStr = prevStr;
            } else {
                currStr.append(c);
            }
        }

        return currStr.toString();
    }
}
